pkg/virt-controller/services/template.go

```diff
@@ -845,20 +841,21 @@ func (t *templateService) RenderHotplugAttachmentPodTemplate(volumes []*v1.Volum
 				},
 			},
 		})
-		if !skipMount {
-			pvc := claimMap[volume.Name]
-			if pvc != nil {
-				if types.IsPVCBlock(pvc.Spec.VolumeMode) {
-					pod.Spec.Containers[0].VolumeDevices = append(pod.Spec.Containers[0].VolumeDevices, k8sv1.VolumeDevice{
-						Name:       volume.Name,
-						DevicePath: fmt.Sprintf("/path/%s/%s", volume.Name, pvc.GetUID()),
-					})
-				} else {
-					pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts, k8sv1.VolumeMount{
-						Name:      volume.Name,
-						MountPath: fmt.Sprintf("/%s", volume.Name),
-					})
-				}
+		pvc := claimMap[volume.Name]
+		if pvc == nil {
+			continue
+		}
+		if types.IsPVCBlock(pvc.Spec.VolumeMode) {
+			pod.Spec.Containers[0].VolumeDevices = append(pod.Spec.Containers[0].VolumeDevices, k8sv1.VolumeDevice{
+				Name:       volume.Name,
+				DevicePath: fmt.Sprintf("/path/%s/%s", volume.Name, pvc.GetUID()),
+			})
+		} else {
+			if !skipMount {
```

Comment:
I remember that at the times when hotplug volumes with a single attachment pod and live migration support were introduced, we also had some conversations with @awels regarding `VolumeMounts/VolumeDevices`. But as of now I cannot recall the details why it was decided to not add the already attached volumes. Perhaps it was related to issues with some specific CSI... Anyway, with a single attachment pod and current implementation, I cannot think of any particular problem that may arise if we mount the volume here. When adding new volumes, we first delete the existing pod and then create a new one... so I guess it should not cause troubles with making it available for multiple pods.

However, if doing it for block volumes, then I suppose the same behavior should also be used for filesystem. Or is there any reason why we would want hotplug volumes to behave differently depending on the type?

Actually, a similar approach was suggested in https://github.com/kubevirt/kubevirt/pull/6728#issuecomment-1081062069. But seems it never moved from there.

Ping @awels, @alicefr, @mhenriks, Maybe you guys have thoughts here?

Reply:
Thanks, @vasiliy-ul. I checked the https://github.com/kubevirt/kubevirt/pull/6728#issuecomment-1081062069 you mentioned. Yes, that is a similar approach. I do not know more details about this skipping, so I only change a few works (as you see, I do not change the FS-related behavior). I describe more detail with my changes in the following https://github.com/kubevirt/kubevirt/pull/9269#issuecomment-1436764795.

Thanks for your help here!

Reply:
> Thanks, @vasiliy-ul. I checked the [#6728 (comment)](https://github.com/kubevirt/kubevirt/pull/6728#issuecomment-1081062069) you mentioned. Yes, that is a similar approach. I do not know more details about this skipping, so I only change a few works (as you see, I do not change the FS-related behavior). I describe more detail with my changes in the following [#9269 (comment)](https://github.com/kubevirt/kubevirt/pull/9269#issuecomment-1436764795).
> 
> Thanks for your help here!

You aren't changing the FS but the block devices yes. I'm not sure if we can always assign a block device to the hotplug pod. This might work for ISCSI csi protocol but not for other storage providers. Before accepting this change, we need to be sure about the behavior of the block devices

Reply:
While going through the previous discussions about this topic, I've got an impression that the main problem here is that CSI implementations behave differently. Some of them do not support attaching RWO to several pods, hence we need to first delete the existing attachment pod and then create a new one. Some others (like in this case) on the opposite does not like such a flow and requires that a new attachment pod is created before the old one gets dropped. I wonder: can we support both use cases in KubeVirt? Like by introducing a policy setting in the CR (e.g. `attachmentPodCreationPolicy`) which will allow configuring the creation/deletion order of attachment pods?

Reply:
@vasiliy-ul how would you know when you need to assign this policy? Generally, kubevirt doesn't introspect which kind of storage class is the PVC. In this case, you would also need some kind of mapping that recognize if that storage class conflict with the default hotplug attachment policy

Reply:
I think the policy should be adjusted by the end-user. Not proposing to keep a mapping of storage classes. Probably the default behavior would be to keep the existing flow (i.e. delete-and-create). But when needed, this can be changed to create-and-delete by modifying the setting in the CR.

Reply:
... ah, you mean a mapping to handle the case when there are multiple storage classes in the cluster from different providers. And how to assign individual policy to each of them, right?

Reply:
yes, such a policy depends on the storage class of the PVC and you might have different storage providers with different behaviors. It is a tricky problem to solve. 
For example, certain storage providers implement differently the RWO mode. In general, this should mean attaching to a single node, but multiple pods on the same node could access the storage. However, in certain cases, this means a single pod. That wasn't one of the reasons why https://github.com/kubevirt/kubevirt/pull/6728 was raising concerns

Reply:
Yeah, that is a valid point. The policy setting in the CR will not solve all the problems. However, it might still be an improvement and a short-term fix for some use cases while elaborating a better solution.

Reply:
I'm not sure if it should be a policy in the CR or rather an annotation set by the user. It depends on who is responsible to know this information.

Reply:
Thanks @alicefr,
I agree that changes might introduce some side-effect on other storage providers. 

For the behavior of block devices on the hot-plug pod. The `kubelet` would call node stage/publish again if the `voluemDevices` contain the item.
For now, we set up the volume in `Volumes` but not in the `VolumeDevices` when rendering the second `hp-volume-` pod.
At this moment, the `kubelet` would think this volume is unused, so it would do no-op. I am confused why we want to make`kubelet` think this volume is unused. (with the second `hp-volume-` pod handling)

Reply:
@Vicente-Cheng I think @awels's comment [here](https://github.com/kubevirt/kubevirt/pull/5649#issuecomment-865913057) might answer your question

Reply:
Thanks, @alicefr. 
That's good tips for the reason that we using the skipMount.

I thought that was the reason (from the comment), right?
> We were having issues with for instance rook-ceph-rbd RWO filesystem volumes, where attempting to mount on the second pod would fail because there was a RW file system mounted somewhere else (in the VM). At first we attempted a read only mount in the flip flop pod, but some provisioners don't support read only volumes (in particular longhorn). But we then realized the real purpose of the attachment pods is to attach the volume to the node and have a reference that can properly clean up the attachment, but we don't really care if the volume is mounted in that pod once the VM is using it.

A quick checking for the rbd CSI behavior. (If I got some misunderstanding, please correct me. Thanks!)
On the NodeStage path, it looks like two following cases:
1. Block volume, use the bind mount
2. FS volume, use mount directly
On the NodePublish path, it looks similar to the above.
1. Block volume, create mount path
2. FS volume, create mount directory

Does the above problem have happened on the FS volume?

Reply:
On this, we need to ask @awels, I don't know if we have encounter the problem on fs volumes

Reply:
Sure, we could wait for his thoughts. Then, he would be able to tell me more about this situation. And indeed, we need to consider more about both fs volumes/block volumes. Thanks!

Reply:
Note I don't know https://github.com/kubevirt/kubevirt/pull/6728 and this PR are solving the same issue. The approach is definitely different. In the other PR we create a new pod first, and then delete the old pod. In this PR the order remains the same, it is just that for block volumes, we go through the CSI publish stage because we are specifying the volumeDevices section in the new pod for existing volumes.

@Vicente-Cheng have you verified this fixes the issue for your provisioner?

Reply:
And when you verify, have you also verified that if you use a filesystem volume on your provisioner that you don't have the exact same issue?